### ISSUE TITLE
Wrong locales path during SSR

### ISSUE THREAD

**Issue by kleinpetr** [s00d/nuxt-i18n-micro#82]
**Describe the bug**
I am facing a Vercel related bug, when transaltion keys are not resolved during SSR, It's a bit complex project, so unfortunately I was not able to make a minimal reproduction, we use over 14 layers

But  I am trying to figure out why locales can't be resolved and it seem, it tries to load locales from a wrong path as you can see from Vercel logs here (we have `disablePageLocales` set to `true` 

![image](https://github.com/user-attachments/assets/d887aa3c-bbc1-468c-8ecc-77ef01bd8c99)

It tries to get data from `/_locales/index/en/data.json` where there are no data, it returns just empty json.

When I check the build output on the vercel I can see only `general` locales, so I also don't know how it can resolves the previous path with empty json.

![image](https://github.com/user-attachments/assets/bef69585-5309-4258-9897-977bae33bdb7)

when I access url with `/_locales/general/en/data.json` the value is there.

in the bundle there is also top level folder called `locales` without underscore, where is the same output, could you explain a bit how are translations resolved both on server and client? :pray: 

Could you give me a clue what I could try to change to get understanding what happens under the hood?
Locally it works fine.

Thank you so much.

**Screenshots**
![image](https://github.com/user-attachments/assets/92b02420-b16c-42dd-b1e1-2331c3db2261)

**Comment by s00d** [comment]
Hi, this is strange behavior. Do you have multiple layers? Maybe disablePageLocales isn’t enabled in one of them? It shouldn’t be fetching locales for the index if disablePageLocales is enabled.

I tested it here and didn’t see this issue:
https://github.com/s00d/-Cloudflare-Pages-project/tree/vercel

**Comment by kleinpetr** [comment]
it's weird. We have lot of layers but I would expect
- the configuration will be inherited from the bottom to top
- in case where page locales would be eanbled for some reason, general locales will be resolved as well

is there any idea how I could investigate it? because the most strange thing is, that locally everything works fine.

**Comment by kleinpetr** [comment]
even though I set this flag explicitly in every layer, the problem stays same :thinking:

**Comment by s00d** [comment]
Unfortunately, I can only guess, and I don't understand what the problem might be. I need a live example to figure it out; without code, nothing is clear.

**Comment by kleinpetr** [comment]
Ok, yesterday I spent about an hour with @danielroe to explore the problem, unfortunately without a solution, we figured out a small bug where page locales endpoint is called even though the `disablePageLocales` is set to true.

but I played a lot around it today, and tried to patch the plugin and add some console logs in order to see what is happening during load translations on the server and client and what is different. So this is what I have now

patched `translationHelper`
```js
const generalLocaleCache = {};
const routeLocaleCache = {};
const dynamicTranslationsCaches = [];
const serverTranslationCache = {};
const serverTranslationInit = {};
function deepClone(value) {
  if (Array.isArray(value)) {
    return JSON.parse(JSON.stringify(value));
  } else if (typeof value === "object" && value !== null) {
    return JSON.parse(JSON.stringify(value));
  }
  return value;
}
function findTranslation(translations, parts) {
  console.warn(`findTranslation() -> parts: ${parts} | translations:`, translations)
  let value = translations;
  for (const part of parts) {
    if (value && typeof value === "object" && part in value) {
      console.warn(`findTranslation(${parts}) -> part value`)
      value = value[part];
    } else {
      console.warn(`findTranslation(${parts}) -> null value`)
      return null;
    }
  }
  if (typeof value === "object" && value !== null) {
    console.warn(`findTranslation(${parts}) -> deepClone value, translations: ${translations}`)
    return deepClone(value);
  }
  return value ?? null;
}
export function useTranslationHelper() {
  return {
    hasCache(locale, page) {
      return (serverTranslationCache[`${locale}:${page}`] ?? /* @__PURE__ */ new Map()).size > 0;
    },
    getCache(locale, routeName) {
      return serverTranslationCache[`${locale}:${routeName}`];
    },
    setCache(locale, routeName, cache) {
      const cacheKey = `${locale}:${routeName}`;
      serverTranslationCache[`${locale}:${routeName}`] = cache;
      serverTranslationInit[`${locale}:index`] = true;
      serverTranslationInit[cacheKey] = true;
    },
    mergeTranslation(locale, routeName, newTranslations, force = false) {
      if (!force && !routeLocaleCache[`${locale}:${routeName}`]) {
        console.error(`marge: route ${locale}:${routeName} not loaded`);
      }
      routeLocaleCache[`${locale}:${routeName}`] = {
        ...routeLocaleCache[`${locale}:${routeName}`],
        ...newTranslations
      };
    },
    mergeGlobalTranslation(locale, newTranslations, force = false) {
      if (!force && !generalLocaleCache[`${locale}`]) {
        console.error(`marge: route ${locale} not loaded`);
      }
      generalLocaleCache[locale] = {
        ...generalLocaleCache[locale],
        ...newTranslations
      };
    },
    hasGeneralTranslation(locale) {
      return !!generalLocaleCache[locale];
    },
    hasPageTranslation(locale, routeName) {
      return !!routeLocaleCache[`${locale}:${routeName}`];
    },
    getTranslation: (locale, routeName, key) => {
      const cacheKey = `${locale}:${routeName}`;
      const cached = serverTranslationCache[cacheKey]?.get(key);
      if (cached) {
        console.warn('i18nHelper.getTranslation() -> cached value:', cached)
        return cached;
      }
      const parts = key.split(".");
      let result = null;
      for (const dynamicCache of dynamicTranslationsCaches) {
        console.warn('i18nHelper.getTranslation() -> dynamicCache:', dynamicCache)
        result = findTranslation(dynamicCache[locale] || null, parts);
        if (result !== null) break;
      }
      if (!result) {
        console.warn('i18nHelper.getTranslation() -> routeLocaleCache:', routeLocaleCache)
        console.warn('i18nHelper.getTranslation() -> generalLocaleCache:', generalLocaleCache)
        result = findTranslation(routeLocaleCache[cacheKey] || null, parts) ?? findTranslation(generalLocaleCache[locale] || null, parts);
      }
      if (result) {
        if (!serverTranslationCache[cacheKey]) {
          serverTranslationCache[cacheKey] = /* @__PURE__ */ new Map();
        }
        serverTranslationCache[cacheKey].set(key, result);
      }
      return result;
    },
    async loadPageTranslations(locale, routeName, translations) {
      const cacheKey = `${locale}:${routeName}`;
      routeLocaleCache[cacheKey] = { ...translations };
      serverTranslationInit[cacheKey] = true;
    },
    async loadTranslations(locale, translations) {
      generalLocaleCache[locale] = { ...translations };
      serverTranslationInit[`${locale}:index`] = true;
    }
  };
}
```

This is the SSR output
![image](https://github.com/user-attachments/assets/1323e418-d6f7-4ddf-8d8f-01e04ac55eee)

This is a client output
![image](https://github.com/user-attachments/assets/408b0bea-21d7-4c7e-9f7a-c10d49938784)

**Comment by kleinpetr** [comment]
Probably I got a function which does this problem, and it's `hasGeneralTranslations` I've put some logs there

```js
hasGeneralTranslation(locale) {
      console.warn(`hasGeneralTranslation(${locale}) -> generalLocaleCache:`, generalLocaleCache)
      console.warn('!!generalLocaleCache[locale]:', !!generalLocaleCache[locale])
      return !!generalLocaleCache[locale];
    },
```

and the output is

SSR
![image](https://github.com/user-attachments/assets/22b5b0b3-85a5-4310-9631-6023bafe60f6)

Client
![image](https://github.com/user-attachments/assets/a1232df4-0186-4e3b-95af-e650fbf02464)


I am not sure where the `generalLocalCache` should come from since I found just `mergeGlobalTranslations()` which works with this property, where I also added some logs, but unfortunately it seems that it is not called on neither ssr nor client.

I hope this helps you guess where the problem might be coming from. :pray:

**Comment by kleinpetr** [comment]
I was try to enforce this condition to go to the next step, which is `loadTranslations` where I debug a path and data from the `$fetch` itself. Paths are same but on the SSR there are no data at all.. any idea what is going on?

SSR
![image](https://github.com/user-attachments/assets/4ee784e2-d57e-4601-8141-442e98df3357)


Client
![image](https://github.com/user-attachments/assets/fcac1b79-4c1f-4754-b372-ed8db3cc9e79)

**Comment by s00d** [comment]
Hi, maybe you just need to update it? The issue might be related to [this](https://github.com/s00d/nuxt-i18n-micro/pull/87)?

**Comment by s00d** [comment]
Hi!

If you can create a simple project where this issue can be reproduced, I’ll most likely be able to fix it. Right now, I can’t make sense of the problem based on the information you’ve provided. From what I see, your server seems unable to fetch translations from the cache. This happens here:  
[**`i18n-loader.ts`**](https://github.com/s00d/nuxt-i18n-micro/blob/main/src/runtime/server/middleware/i18n-loader.ts).

This file is responsible for loading translations and caching them. It uses a two-level cache:  
1. **File-based cache**  (this is specifically useful for environments like Vercel).  
2. **In-memory cache**

From the logs, it looks like translations are simply not accessible on your server. However, I’ve personally tested this multiple times using these links:  
- [Test Project on GitHub](https://github.com/s00d/-Cloudflare-Pages-project/tree/vercel)  
- [Deployed Demo](https://cloudflare-pages-project.vercel.app/en)  

The issue could stem from anything, such as:  
- Dependency versions  
- Nuxt configuration  
- Dependency conflicts  
- Vercel settings  
- A bug under specific conditions  
- Or something else entirely.  

In short, without a test project to reproduce the issue, I can’t pinpoint the problem.

**Comment by kleinpetr** [comment]
I am wondering if it could does the $fetch itself 🤔  maybe trying to resolve the endpoint locally from registered routes, but I guess that the library adds _locales to the root of the build so I think it is not included in the Nuxt/Nitro app itself. but maybe I am wrong

**Comment by danielroe** [comment]
On serverless platforms, `$fetch` won't be able to fetch static files as they are stored on the CDN and not internally to the nitro server. You might be able to set `serveStatic` to work around this temporarily but instead this module should probably implement a way to get this data on the server side using Nitro storage.

**Comment by kleinpetr** [comment]
Okay, so if I add `baseUrl` with the actual origin, not just `/`, it gets the data. So I'm just wondering if there's any way to tell `$fetch` to make an http call and try not to handle it internally, the docs say that the internal request is made by unenv, so I'm not sure if this behavior can be changed by some option in $fetch itself. What do you think @danielroe ?

![image](https://github.com/user-attachments/assets/eedc921d-f255-45f4-9885-bb5bcd2772ae)

**Comment by s00d** [comment]
Try simply setting the apiBaseUrl with the domain (https://example.com/_locales), I'm not sure if it will work like this, I haven't checked.

**Comment by kleinpetr** [comment]
Yes it wokrs this way.

**Comment by s00d** [comment]
I should also warn that the entire Nuxt 3.14.* version is buggy and slow, it's better to use 3.13.* for now.

https://github.com/nuxt/nuxt/issues/30290
https://github.com/nuxt/nuxt/issues/30008

**Comment by kleinpetr** [comment]
on this project we use 
`nuxt: "^3.12.2"`

**Comment by kleinpetr** [comment]
@s00d do you think it's possible to switch to Nitro Storage as Daniel mentioned? :thinking:

**Comment by danielroe** [comment]
@s00d neither of those issues (now resolved) is related to the bug reported here

**Comment by s00d** [comment]
> @s00d do you think it's possible to switch to Nitro Storage as Daniel mentioned? 🤔

If you're referring to https://nitro.build/guide/storage, it is already being used to fetch translations. If you want to fully switch to storage instead of querying through $fetch, in theory, it's a good solution, but unfortunately, I don't have time for this. Perhaps you could do it yourself and submit a pull request.

**Comment by s00d** [comment]
> Nuxt 3.14.*

This error is unrelated, I just wanted to warn you. The problem lies in the Nuxt 3.14.* version itself. After updating to 3.14.*, the project becomes twice as slow and consumes three times more CPU.

**Comment by danielroe** [comment]
@s00d I'm aware - I resolved those issues as I mentioned above; v3.15 is slated for a release this weekend. Not sure what you are trying to tell me.

**Comment by kleinpetr** [comment]
@s00d when I set baseUrl to absolute one, with origin, I am getting this error during prerendering..

![image](https://github.com/user-attachments/assets/c6124d91-781f-4944-9381-1bf748219578)

**Comment by s00d** [comment]
> @s00d when I set baseUrl to absolute one, with origin, I am getting this error during prerendering..
> 
> ![image](https://private-user-images.githubusercontent.com/17269309/399590688-c6124d91-781f-4944-9381-1bf748219578.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU3NzEwMTgsIm5iZiI6MTczNTc3MDcxOCwicGF0aCI6Ii8xNzI2OTMwOS8zOTk1OTA2ODgtYzYxMjRkOTEtNzgxZi00OTQ0LTkzODEtMWJmNzQ4MjE5NTc4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTAxVDIyMzE1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4MGJhMjIxOWI1MTA5MzExNjhlODlmYjEzODdlNDgyNDRhYWVhNDVjNjUyMTc2ZTAxMDgwYzBjMWM5MDNlNDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sqwkFhuZTazbQFONSfsh-Q1L_ion1iv0W8SxyiPP_B4)

Hi, please create examples with an error, I can't figure out what the problem is from the screenshot.

**Comment by danielroe** [comment]
@s00d would you value a PR implementing the storage solution I suggest?

**Comment by s00d** [comment]
> @s00d would you value a PR implementing the storage solution I suggest?

Yes, of course, I’d be happy to review and accept your PR!

**Comment by kleinpetr** [comment]
@s00d finally.. I don't know how it's possible, but after switching from 1.60 to 1.64 the problem came back again...

**Comment by s00d** [comment]
> [@s00d](https://github.com/s00d) finally.. I don't know how it's possible, but after switching from 1.60 to 1.64 the problem came back again...

Try version 1.68.0. I rolled back the server logic to the previous version. The issues started after this pull request:  
https://github.com/s00d/nuxt-i18n-micro/issues/101  

The task seemed interesting, but the implementation fell short. I couldn't find examples or documentation on working with the Nitro store. Other modules also seem strange—it's as if no one actually uses the Nitro store for tasks like this. For example, Content has its own database, even though in earlier versions I noticed they tried to integrate it.  

In any case, everything should be working again now.
